packages/react-query/src/useBaseQuery.ts

```diff
@@ -131,6 +136,26 @@ export function useBaseQuery<
     result,
   )
 
+  if (
+    defaultedOptions.experimental_promise &&
+    !isServer &&
+    willFetch(result, isRestoring) &&
+    !observer.hasListeners()
+  ) {
+    const promise = isNewCacheEntry
+      ? // Fetch immediately on mount in order to ensure `.promise` is resolved even if the component is unmounted
+        fetchOptimistic(defaultedOptions, observer, errorResetBoundary)
+      : // subscribe to the "cache promise" so that we can finalize the currentThenable once data comes in
+        client.getQueryCache().get(defaultedOptions.queryHash)?.promise
+
+    promise?.catch(noop).finally(() => {
+      if (!observer.hasListeners()) {
```

Comment:
I'm trying to wrap my head around this all. Why are we checking for `!observer.hasListeners()` here?

Reply:
Just to prevent `updateResult()` to be called twice - if it has listeners, `updateResult` will be called

Reply:
yeah I just checked, if we remove both checks for `!observer.hasListeners()`, all the tests still pass. if we want to keep these checks, we should add tests where this fails @KATT. I know that calling `updateResult` multiple times might e.g. re-run inline `select` functions, but I'm not sure if that's worth optimizing for ...

Reply:
I tried adding a test for it by

```ts
vi.spyOn(QueryObserver.prototype, 'updateResult')
```

but it's called a bunch of times within a render so adding a test the way I thought will be quite hard

feel free to delete the optimization

Reply:
I've deleted the first check, because it's technically impossible that `willFetch` is true _and_ we have an observer. I've kept the second check because maybe an observer shows up in between: [a9148ba](https://github.com/TanStack/query/pull/7988/commits/a9148baaa678fcb5186b985c87e797d7a43f04dd)